File: app/src/main/java/arun/com/chromer/activities/browsing/article/ChromerArticleActivity.java
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.activities.browsing.article;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.LocalBroadcastManager;
import android.view.Menu;
import android.view.MenuItem;

import com.mikepenz.community_material_typeface_library.CommunityMaterial;
import com.mikepenz.iconics.IconicsDrawable;

import arun.com.chromer.R;
import arun.com.chromer.activities.CustomTabActivity;
import arun.com.chromer.activities.MoreMenuActivity;
import arun.com.chromer.activities.OpenIntentWithActivity;
import arun.com.chromer.activities.settings.Preferences;
import arun.com.chromer.customtabs.callbacks.ClipboardService;
import arun.com.chromer.customtabs.callbacks.FavShareBroadcastReceiver;
import arun.com.chromer.customtabs.callbacks.SecondaryBrowserReceiver;
import arun.com.chromer.data.history.HistoryRepository;
import arun.com.chromer.data.website.model.WebSite;
import arun.com.chromer.util.RxUtils;
import arun.com.chromer.util.Utils;
import timber.log.Timber;
import xyz.klinker.android.article.ArticleActivity;
import xyz.klinker.android.article.data.webarticle.model.WebArticle;

import static android.content.Intent.FLAG_ACTIVITY_MULTIPLE_TASK;
import static android.graphics.Color.WHITE;
import static arun.com.chromer.activities.settings.Preferences.PREFERRED_ACTION_BROWSER;
import static arun.com.chromer.activities.settings.Preferences.PREFERRED_ACTION_FAV_SHARE;
import static arun.com.chromer.activities.settings.Preferences.PREFERRED_ACTION_GEN_SHARE;
import static arun.com.chromer.shared.Constants.ACTION_MINIMIZE;
import static arun.com.chromer.shared.Constants.EXTRA_KEY_FROM_ARTICLE;
import static arun.com.chromer.shared.Constants.EXTRA_KEY_ORIGINAL_URL;

public class ChromerArticleActivity extends ArticleActivity {
    private String baseUrl = "";
    private BroadcastReceiver minimizeReceiver;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        baseUrl = getIntent().getDataString();
        registerMinimizeReceiver();
    }

    @Override
    protected void onArticleLoadingFailed(Throwable throwable) {
        super.onArticleLoadingFailed(throwable);
        // Start custom tab activity on the same task, silently killing this activity.
        final Intent customTabActivity = new Intent(this, CustomTabActivity.class);
        customTabActivity.setData(getIntent().getData());
        startActivity(customTabActivity);
        finish();
    }

    @Override
    protected void onArticleLoaded(@NonNull WebArticle webArticle) {
        super.onArticleLoaded(webArticle);
        HistoryRepository.getInstance(this)
                .insert(WebSite.fromArticle(webArticle))
                .compose(RxUtils.applySchedulers())
                .subscribe();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.article_view_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        final MenuItem actionButton = menu.findItem(R.id.menu_action_button);
        switch (Preferences.get(this).preferredAction()) {
            case PREFERRED_ACTION_BROWSER:
                final String browser = Preferences.get(this).secondaryBrowserPackage();
                if (Utils.isPackageInstalled(this, browser)) {
                    actionButton.setTitle(R.string.choose_secondary_browser);
                    final ComponentName componentName = ComponentName.unflattenFromString(Preferences.get(this).secondaryBrowserComponent());
                    try {
                        actionButton.setIcon(getPackageManager().getActivityIcon(componentName));
                    } catch (PackageManager.NameNotFoundException e) {
                        actionButton.setVisible(false);
                    }
                } else {
                    actionButton.setVisible(false);
                }
                break;
            case PREFERRED_ACTION_FAV_SHARE:
                final String favShare = Preferences.get(this).favSharePackage();
                if (Utils.isPackageInstalled(this, favShare)) {
                    actionButton.setTitle(R.string.fav_share_app);
                    final ComponentName componentName = ComponentName.unflattenFromString(Preferences.get(this).favShareComponent());
                    try {
                        actionButton.setIcon(getPackageManager().getActivityIcon(componentName));
                    } catch (PackageManager.NameNotFoundException e) {
                        actionButton.setVisible(false);
                    }
                } else {
                    actionButton.setVisible(false);
                }
                break;
            case PREFERRED_ACTION_GEN_SHARE:
                actionButton.setIcon(new IconicsDrawable(this)
                        .icon(CommunityMaterial.Icon.cmd_share_variant)
                        .color(WHITE)
                        .sizeDp(24));
                actionButton.setTitle(R.string.share);
                break;
        }
        final MenuItem favoriteShare = menu.findItem(R.id.menu_share_with);
        final String pkg = Preferences.get(this).favSharePackage();
        if (pkg != null) {
            final String app = Utils.getAppNameWithPackage(this, pkg);
            final String label = String.format(getString(R.string.share_with), app);
            favoriteShare.setTitle(label);
        } else {
            favoriteShare.setVisible(false);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        switch (item.getItemId()) {
            case R.id.menu_action_button:
                switch (Preferences.get(this).preferredAction()) {
                    case PREFERRED_ACTION_BROWSER:
                        sendBroadcast(new Intent(this, SecondaryBrowserReceiver.class).setData(Uri.parse(baseUrl)));
                        break;
                    case PREFERRED_ACTION_FAV_SHARE:
                        sendBroadcast(new Intent(this, FavShareBroadcastReceiver.class).setData(Uri.parse(baseUrl)));
                        break;
                    case PREFERRED_ACTION_GEN_SHARE:
                        shareUrl();
                        break;
                }
                break;
            case R.id.menu_copy_link:
                startService(new Intent(this, ClipboardService.class).setData(Uri.parse(baseUrl)));
                break;
            case R.id.menu_open_with:
                startActivity(new Intent(this, OpenIntentWithActivity.class).setData(Uri.parse(baseUrl)));
                break;
            case R.id.menu_share:
                shareUrl();
                break;
            case R.id.menu_open_full_page:
                final Intent customTabActivity = new Intent(this, CustomTabActivity.class);
                customTabActivity.setData(Uri.parse(baseUrl));
                if (Preferences.get(this).mergeTabs()) {
                    customTabActivity.addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT);
                    customTabActivity.addFlags(FLAG_ACTIVITY_MULTIPLE_TASK);
                }
                startActivity(customTabActivity);
                finish();
                break;
            case R.id.menu_more:
                final Intent moreMenuActivity = new Intent(this, MoreMenuActivity.class);
                moreMenuActivity.setData(getIntent().getData());
                moreMenuActivity.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                moreMenuActivity.putExtra(EXTRA_KEY_ORIGINAL_URL, baseUrl);
                moreMenuActivity.putExtra(EXTRA_KEY_FROM_ARTICLE, true);
                startActivity(moreMenuActivity);
                break;
            case R.id.menu_share_with:
                sendBroadcast(new Intent(this, FavShareBroadcastReceiver.class).setData(Uri.parse(baseUrl)));
                break;
        }
        return true;
    }

    private void shareUrl() {
        Utils.shareText(this, baseUrl);
    }

    private void registerMinimizeReceiver() {
        minimizeReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (intent.getAction().equalsIgnoreCase(ACTION_MINIMIZE) && intent.hasExtra(EXTRA_KEY_ORIGINAL_URL)) {
                    final String url = intent.getStringExtra(EXTRA_KEY_ORIGINAL_URL);
                    if (baseUrl.equalsIgnoreCase(url)) {
                        try {
                            Timber.d("Minimized %s", url);
                            moveTaskToBack(true);
                        } catch (Exception e) {
                            Timber.e(e);
                        }
                    }
                }
            }
        };
        LocalBroadcastManager.getInstance(this).registerReceiver(minimizeReceiver, new IntentFilter(ACTION_MINIMIZE));
    }

    @Override
    protected void onDestroy() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(minimizeReceiver);
        super.onDestroy();
    }
}

File: app/src/main/java/arun/com/chromer/browsing/article/ChromerArticleActivity.kt
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.browsing.article

import android.content.ComponentName
import android.content.Intent
import android.content.Intent.*
import android.content.pm.PackageManager
import android.graphics.Color.WHITE
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import arun.com.chromer.R
import arun.com.chromer.browsing.customtabs.CustomTabActivity
import arun.com.chromer.browsing.customtabs.callbacks.ClipboardService
import arun.com.chromer.browsing.customtabs.callbacks.FavShareBroadcastReceiver
import arun.com.chromer.browsing.customtabs.callbacks.SecondaryBrowserReceiver
import arun.com.chromer.browsing.openwith.OpenIntentWithActivity
import arun.com.chromer.browsing.optionspopup.ChromerOptionsActivity
import arun.com.chromer.data.history.DefaultHistoryRepository
import arun.com.chromer.data.webarticle.model.WebArticle
import arun.com.chromer.data.website.model.Website
import arun.com.chromer.di.activity.ActivityComponent
import arun.com.chromer.settings.Preferences
import arun.com.chromer.settings.Preferences.*
import arun.com.chromer.shared.Constants.EXTRA_KEY_FROM_ARTICLE
import arun.com.chromer.shared.Constants.EXTRA_KEY_ORIGINAL_URL
import arun.com.chromer.tabs.DefaultTabsManager
import arun.com.chromer.util.SchedulerProvider
import arun.com.chromer.util.Utils
import com.mikepenz.community_material_typeface_library.CommunityMaterial
import com.mikepenz.iconics.IconicsDrawable
import javax.inject.Inject

class ChromerArticleActivity : BaseArticleActivity() {
    private var baseUrl: String? = ""

    @Inject
    lateinit var historyRepository: DefaultHistoryRepository

    @Inject
    lateinit var tabsManager: DefaultTabsManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        baseUrl = intent.dataString
    }

    override fun inject(activityComponent: ActivityComponent) {
        activityComponent.inject(this)
    }

    override fun onArticleLoadingFailed(throwable: Throwable?) {
        super.onArticleLoadingFailed(throwable)
        // Loading failed, try to go back to normal url tab if it exists, else start a new normal
        // rendering tab.
        loadInNormalTab()
    }

    override fun onArticleLoaded(webArticle: WebArticle) {
        super.onArticleLoaded(webArticle)
        historyRepository.insert(Website.fromArticle(webArticle))
                .compose(SchedulerProvider.applySchedulers())
                .subscribe()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.article_view_menu, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val actionButton = menu.findItem(R.id.menu_action_button)
        when (Preferences.get(this).preferredAction()) {
            PREFERRED_ACTION_BROWSER -> {
                val browser = Preferences.get(this).secondaryBrowserPackage()
                if (Utils.isPackageInstalled(this, browser)) {
                    actionButton.setTitle(R.string.choose_secondary_browser)
                    val componentName = ComponentName.unflattenFromString(Preferences.get(this).secondaryBrowserComponent()!!)
                    try {
                        actionButton.icon = packageManager.getActivityIcon(componentName)
                    } catch (e: PackageManager.NameNotFoundException) {
                        actionButton.isVisible = false
                    }

                } else {
                    actionButton.isVisible = false
                }
            }
            PREFERRED_ACTION_FAV_SHARE -> {
                val favShare = Preferences.get(this).favSharePackage()
                if (Utils.isPackageInstalled(this, favShare)) {
                    actionButton.setTitle(R.string.fav_share_app)
                    val componentName = ComponentName.unflattenFromString(Preferences.get(this).favShareComponent()!!)
                    try {
                        actionButton.icon = packageManager.getActivityIcon(componentName)
                    } catch (e: PackageManager.NameNotFoundException) {
                        actionButton.isVisible = false
                    }

                } else {
                    actionButton.isVisible = false
                }
            }
            PREFERRED_ACTION_GEN_SHARE -> {
                actionButton.icon = IconicsDrawable(this)
                        .icon(CommunityMaterial.Icon.cmd_share_variant)
                        .color(WHITE)
                        .sizeDp(24)
                actionButton.setTitle(R.string.share)
            }
        }
        val favoriteShare = menu.findItem(R.id.menu_share_with)
        val pkg = Preferences.get(this).favSharePackage()
        if (pkg != null) {
            val app = Utils.getAppNameWithPackage(this, pkg)
            val label = String.format(getString(R.string.share_with), app)
            favoriteShare.title = label
        } else {
            favoriteShare.isVisible = false
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        super.onOptionsItemSelected(item)
        when (item.itemId) {
            R.id.menu_action_button -> when (Preferences.get(this).preferredAction()) {
                PREFERRED_ACTION_BROWSER -> sendBroadcast(Intent(this, SecondaryBrowserReceiver::class.java).setData(Uri.parse(baseUrl)))
                PREFERRED_ACTION_FAV_SHARE -> sendBroadcast(Intent(this, FavShareBroadcastReceiver::class.java).setData(Uri.parse(baseUrl)))
                PREFERRED_ACTION_GEN_SHARE -> shareUrl()
            }
            R.id.menu_copy_link -> startService(Intent(this, ClipboardService::class.java).setData(Uri.parse(baseUrl)))
            R.id.menu_open_with -> startActivity(Intent(this, OpenIntentWithActivity::class.java).setData(Uri.parse(baseUrl)))
            R.id.menu_share -> shareUrl()
            R.id.menu_open_full_page -> loadInNormalTab()
            R.id.menu_more -> {
                val moreMenuActivity = Intent(this, ChromerOptionsActivity::class.java)
                        .apply {
                            data = intent.data
                            putExtra(EXTRA_KEY_ORIGINAL_URL, baseUrl)
                            putExtra(EXTRA_KEY_FROM_ARTICLE, true)
                            addFlags(FLAG_ACTIVITY_NEW_TASK)
                            addFlags(FLAG_ACTIVITY_NEW_DOCUMENT)
                            addFlags(FLAG_ACTIVITY_MULTIPLE_TASK)
                        }
                startActivity(moreMenuActivity)
            }
            R.id.menu_share_with -> sendBroadcast(Intent(this, FavShareBroadcastReceiver::class.java).setData(Uri.parse(baseUrl)))
        }
        return true
    }

    private fun loadInNormalTab() {
        finish()
        if (intent != null && intent.dataString != null) {
            tabsManager.openBrowsingTab(this, Website(intent.dataString!!), true, false, CustomTabActivity::class.java.name)
        }
    }

    private fun shareUrl() {
        Utils.shareText(this, baseUrl)
    }
}
